### PR TITLE
refactor: enable file meta data cache

### LINKED ISSUE #7353: we have a BloomIndexMetaCache, which caches `FileMetaData`, but is dedicated to bloom filter index files now.
we have a BloomIndexMetaCache, which caches `FileMetaData`, but is dedicated to bloom filter index files now.

https://github.com/datafuselabs/databend/blob/1b576fbe78552ca5d024205ee356af70b5b9edf1/src/query/storages/fuse-meta/src/caches/cache.rs#L41.

https://github.com/datafuselabs/databend/blob/26c74149b6b97b7152288b41ea0564114d617aaf/src/query/storages/fuse-meta/src/caches/memory_cache.rs#L50

maybe we could do the refactor in another dedicated PR?

_Originally posted by @dantengsky in https://github.com/datafuselabs/databend/pull/7311#discussion_r956983132_

### PR DESCRIPTION
**Pull request by dantengsky**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

- (re) introduce file metadata cache
- use file metadata cache in hive table
- refactor the file metadata loading parts of bloom filter index loading
   - use cached reader, instead of interacting with cache layer directly
   - respect the length of index file size, so that seek operation no longer need extra IOs

Fixes #7353

### CODE REVIEW

**Comment by Xuanwo** [comment]
Is it possible to port to OpenDAL instead?

**Comment by sandflee** [comment]
hive sql may read tens of thousand of files, so is the cache has upper limit?  how to erase the item when cache is full? TTL based or LRU?

**Comment by dantengsky** [comment]
> hive sql may read tens of thousand of files, so is the cache has upper limit? 

the upper limit now is set to 3000 (hard coded)

https://github.com/datafuselabs/databend/blob/49a0349a4667001931620f22cdc83e72a7bbd359/src/query/storages/fuse-meta/src/caches/cache.rs#L68-L74


> how to erase the item when cache is full? TTL based or LRU?

LRU

**Comment by dantengsky** [comment]
> Is it possible to port to OpenDAL instead? Maybe as a layer?

I am not sure... currently, the items being cached are mostly of objects, instead of raw bytes though

**Comment by Xuanwo** [comment]
> I am not sure... currently, the items being cached are mostly of objects, instead of raw bytes though

I made a mistake. This PR intends to cache the `FileMetadata` of parquet, which differs from OpenDAL's ObjectMetadata.

For OpenDAL's ObjectMetadata, we can implement https://github.com/datafuselabs/opendal/issues/605

File: src/query/storages/fuse-meta/src/caches/cache.rs [hunk 1]
```diff
@@ -1,5 +1,3 @@
-// Copyright 2021 Datafuse Labs.
```
Comment (Xuanwo) [inline src/query/storages/fuse-meta/src/caches/cache.rs:1]: Why removed?

Comment (dantengsky) [reply src/query/storages/fuse-meta/src/caches/cache.rs:1]: oops, that is a typo, let me add it back

**Review by zhyass — APPROVED** [review]
LGTM

**Comment by dantengsky** [comment]
@Xuanwo  PTAL

**Review by Xuanwo — APPROVED** [review]
Great work!

**Pull request merged**